File: LeetCode/211. Design Add and Search Words Data Structure/Solution.js
var WordDictionary = function () {
  this.root = {};
};

/**
 * @param {string} word
 * @return {void}
 */
WordDictionary.prototype.addWord = function (word) {
  let node = this.root;

  for (let chr of word) {
    if (chr in node) node = node[chr];
    else {
      node[chr] = {};
      node = node[chr];
    }
  }

  node.end = true;
};

/**
 * @param {string} word
 * @return {boolean}
 */
WordDictionary.prototype.search = function (word) {
  return find(this.root, word, 0, word.length);

  function find(node, word, pos, len) {
    if (!node) return false;
    if (pos == len) return node.end === undefined ? false : node.end;

    if (word[pos] === ".") {
      for (let chr in node) {
        if (chr == "end") continue;
        if (find(node[chr], word, pos + 1, len)) return true;
      }

      return false;
    } else {
      if (word[pos] in node) return find(node[word[pos]], word, pos + 1, len);
      else return false;
    }
  }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * var obj = new WordDictionary()
 * obj.addWord(word)
 * var param_2 = obj.search(word)
 */

File: LeetCode/211. Design Add and Search Words Data Structure/Solution.py
class WordDictionary:

    def __init__(self):
        self.root = {}
        

    def addWord(self, word: str) -> None:
        cur = self.root
        for char in word:
            if char not in cur : cur[char] = {}
            cur = cur[char]
        cur['end'] = True
    
    def find(self, word: str, node: dict, pos: int, ln: int) -> bool:
        if node is None : return false
        if pos == ln : 
            if 'end' in node:
                return True
            else :
                return False
        
        if word[pos] == '.' :
            for key in node:
                if key != 'end' and self.find(word, node[key], pos + 1, ln):
                    return True
            return False
        
        if word[pos] in node:
            return self.find(word, node[word[pos]], pos + 1, ln)
        else:
            return False
            

    def search(self, word: str) -> bool:
        return self.find(word, self.root, 0, len(word))


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)